hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,244 @@
+              reconstructionCommand);
+        } else {
+          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
+              + "to insufficient source replicas found. Number of source "
+              + "replicas needed: {}. Number of available source replicas are:"
+              + "  {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      } else if (result instanceof ContainerHealthResult
+          .UnderReplicatedHealthResult && ((ContainerHealthResult
+          .UnderReplicatedHealthResult) result)
+          .underReplicatedDueToDecommission()) {
+        Map<DatanodeDetails, SCMCommand> commands = new HashMap<>();
+        Set<Integer> decomIndexes = replicaCount.decommissioningIndexes();
+        final List<DatanodeDetails> selectedDatanodes =
+            getTargetDatanodes(replicas, container, decomIndexes.size());
+        Iterator<DatanodeDetails> iterator = selectedDatanodes.iterator();
+        // In this case we need to do one to one copy.
+        for (ContainerReplica replica : replicas) {
+          if (decomIndexes.contains(replica.getReplicaIndex())) {
+            if (!iterator.hasNext()) {
+              LOG.warn("Couldn't find the enough targets. Source nodes:"
+                      + " {} and the target nodes: {}",
+                  sources, selectedDatanodes);
+              break;
+            }
+            DatanodeDetails decommissioningSrcNode
+                = replica.getDatanodeDetails();
+            final ReplicateContainerCommand replicateCommand =
+                new ReplicateContainerCommand(id.getProtobuf().getId(),
+                    ImmutableList.of(decommissioningSrcNode));
+            DatanodeDetails target = iterator.next();
+            commands.put(target, replicateCommand);
+          }
+        }
+        return commands;
+      } else {
+        if (replicaCount.unRecoverable()) {
+          LOG.warn("The container {} is unrecoverable due to lack of"
+              + " available source replicas.", id);
+        } else {
+          LOG.info("There are no missing indexes or decommissioning indexes"
+                  + " found for {}", id);
+        }
+      }
+
+    } catch (IOException | IllegalStateException ex) {
+      LOG.warn("Exception while processing for creating the EC reconstruction" +
+              " container commands for {}.",
+          id, ex);
+    }
+    return null;
+  }
+
+  private List<DatanodeDetails> getTargetDatanodes(
+      Set<ContainerReplica> replicas, ContainerInfo container,
+      int requiredNodes) throws IOException {
+    // We should ensure that the target datanode has enough space
+    // for a complete container to be created, but since the container
+    // size may be changed smaller than origin, we should be defensive.
+    final long dataSizeRequired =
+        Math.max(container.getUsedBytes(), currentContainerSize);
```

Comment:
according to the logic here, we always have to find a datanode that has at least 5G disk space, even if the container usedBytes is much smaller (for example , 100M). i am not sure if it is reasonable.

Reply:
This container size check I just borrowed from LegacyRM#HandleUnderreplication. I think here, some times we go beyond the container sizes. To make sure we cover that case, otherwise configured container size.